### PR TITLE
i#4549 GA: Configure failure notifs to dynamorio-devs.

### PR DESCRIPTION
**Pull request by abhinav92003**
Uses the dawidd6/action-send-mail@v2 action to send notifications via email to dynamorio-devs for all failures in CI or package workflows.

This is the final piece in the migration to Github Actions.

Fixes: #4549

### CODE REVIEW

**Comment by abhinav92003** [comment]
@derekbruening Will sync offline to add the two new Github secrets (for username and password of the email account) to the DynamoRIO repo settings. 

One caveat of this approach, for PRs from forked repos:
> Secrets are not passed to workflows that are triggered by a pull request from a fork.

I haven't tested behaviour on forked repo PRs yet, but it probably means that we won't get these notification emails for CI failures on those.
But the test suites still run as expected. Note that the Github secret is currently not available in the DynamoRIO repo; but in the above commits, the failed workflows do have complete logs from failed tests, and the successful workflows aren't affected.

File: .github/workflows/ci-osx.yml [hunk 1]
```diff
@@ -78,3 +78,25 @@ jobs:
         # To find available versions, add the following as a step above:
         #  - run: ls -l /Applications
         DEVELOPER_DIR: /Applications/Xcode_11.3.1.app/Contents/Developer
+
+    - name: Send failure mail to dynamorio-devs
```
Comment (abhinav92003) [inline .github/workflows/ci-osx.yml:82]: I don't really like this duplication... but even if we create our own custom action with the common stuff, we'll remove only 7/21 of these lines (will need to keep the ones that access env vars, and the ones required to invoke the new action).

Comment (derekbruening) [reply .github/workflows/ci-osx.yml:82]: We already have a lot of duplication among our 6 workflow files and it's sort of a pain to add something common b/c you have to edit 6 files...but OTOH hopefully we'll not have to touch them again for years and it's prob not worth a lot of effort to share.

**Comment by derekbruening** [comment]
> I haven't tested behaviour on forked repo PRs yet, but it probably means that we won't get these notification emails for CI failures on those.

We want the list emails for pushes to master (not important for PR failures), and I would assume that a forked repo will always be merged by a project member into master and that merge will trigger the email?

File: .github/workflows/ci-aarchxx.yml [hunk 2]
```diff
@@ -75,6 +75,28 @@ jobs:
...
 
+    - name: Send failure mail to dynamorio-devs
+      if: failure()
+      uses: dawidd6/action-send-mail@v2
+      with:
+        server_address: smtp.gmail.com
+        server_port: 465
+        username: ${{secrets.NOTIFICATION_EMAIL_USERNAME}}
+        password: ${{secrets.NOTIFICATION_EMAIL_PASSWORD}}
+        subject: Github Actions CI FAILED on ${{github.repository}} ${{github.ref}}
```
Comment (derekbruening) [inline .github/workflows/ci-aarchxx.yml:86]: I would say include `github.event_name` so we know whether it's `pull request` or whatever the string is for a cronjob at a glance.

Maybe include github.workflow in the title too so we can tell at a glance whether it's osx or windows or what?

File: .github/workflows/ci-aarchxx.yml [hunk 3]
```diff
@@ -75,6 +75,28 @@ jobs:
       env:
         DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY: yes
 
+    - name: Send failure mail to dynamorio-devs
+      if: failure()
```
Comment (derekbruening) [inline .github/workflows/ci-aarchxx.yml:79]: Hmm, in the past we did have list emails on all PR failures.  I guess we can set up filters if we want to ignore PR failures (a reason to put `event_name` in the subject).  Or should we limit emails to cron jobs and master pushes?

File: core/arch/aarchxx/emit_utils.c [hunk 4]
```diff
@@ -50,7 +50,7 @@ insert_load_dr_tls_base(dcontext_t *dcontext, instrlist_t *ilist, instr_t *where
     /* Load TLS base from user-mode thread pointer/ID register:
      * mrs reg_base, tpidr_el0
      */
-    PRE(ilist, where,
+    PRE(ilist, whre,
```
Comment (derekbruening) [inline core/arch/aarchxx/emit_utils.c:53]: This and below are supposed to be temporary do-not-check-in I assume -- I would suggest always having a comment saying that as a safe practice.

**Comment by abhinav92003** [comment]
> We want the list emails for pushes to master (not important for PR failures), and I would assume that a forked repo will always be merged by a project member into master and that merge will trigger the email?

The master merge done by a project member should trigger the email on failure, I believe.

I'm going to try to test our assumptions about forked repo PRs but creating such a PR with i4549-failure-email-notifs as target.

Thanks for adding the GIthub Secrets. Will update the PR when testing is complete and ready for review.

**Comment by abhinav92003** [comment]
Observations from testing for forked repo PRs; see PR #4634 which was forked from abhinav92003/dynamorio and merged into DynamoRIO/dynamorio:i4549-failure-email-notifs for testing:
- Secrets from DynamoRIO/dynamorio are not available to any forked repo PRs. This means that forked repo PRs won't be able to send failure notif emails (see CI logs on PR #4634); unless the forked repo also defines the same Secret variables as DynamoRIO/dynamorio, in which case that other email address will be used. (See other comment thread: This is probably okay as we don't want to send failure notif emails on non-master branches anyway.) We should rename our Secret vars to have a DYNAMORIO_ prefix.

- If CI succeeds, the email notif issue doesn't hinder green-ness of the checks.

- When the PR is merged to a DR branch, it does send failure notif emails (see commit b9344bc above).

File: .github/workflows/ci-aarchxx.yml [hunk 5]
```diff
@@ -75,6 +75,30 @@ jobs:
       env:
         DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY: yes
 
+    - name: Send failure mail to dynamorio-devs
+      if: failure() && github.ref == 'refs/heads/master'
+      uses: dawidd6/action-send-mail@v2
+      with:
+        server_address: smtp.gmail.com
+        server_port: 465
+        username: ${{secrets.DYNAMORIO_NOTIFICATION_EMAIL_USERNAME}}
```
Comment (abhinav92003) [inline .github/workflows/ci-aarchxx.yml:84]: @derekbruening Can you please rename the Secret vars? It is better to have a DYNAMORIO_ prefix on these variables to reduce chances of conflict with a forked repo's Secret vars. See comment on  main PR thread for details.

File: core/arch/aarchxx/emit_utils.c (hunk 4 above)
Comment (abhinav92003) [reply core/arch/aarchxx/emit_utils.c]: Will do in future. Thanks.

Comment (abhinav92003) [reply core/arch/aarchxx/emit_utils.c]: Reverted these temporary changes.

File: .github/workflows/ci-aarchxx.yml (hunk 3 above)
Comment (abhinav92003) [reply .github/workflows/ci-aarchxx.yml]: I've added a condition to send email only for failures on master branch (I assume the cron jobs should also satisfy that condition as they are also on the  master branch). That seems better than folks having to add email filters. Failures on master branch will probably be what others are mostly interested in. And the author of the PR gets emails directly from Github for CI triggered by them.

File: .github/workflows/ci-aarchxx.yml (hunk 2 above)
Comment (abhinav92003) [reply .github/workflows/ci-aarchxx.yml]: Done.

**Comment by abhinav92003** [comment]
Some other observations:
- Commit SHA for the synchronize event (which we use for triggering CI on PR branch pushes) is different from the hash of the commit that triggered the synchronize event.
E.g., following is an email to dynamorio-devs. Note that the SHA doesn't match any commit on PR #4631.
```
Github Actions CI workflow run FAILED!
Workflow: ci-aarchxx/a64-on-x86
Repository: DynamoRIO/dynamorio
Branch ref: refs/pull/4631/merge
SHA: dbb985c6121274ae0e30bda2fc80c0afea062d18
Triggering actor: abhinav92003
Triggering event: pull_request
Run Id: 431475676
See more details on github.com/DynamoRIO/dynamorio/actions/runs/431475676
```

- clang-format doesn't seem to fail after the PR is pushed to master, even though it did fail on the PR. Found this while testing on my forked repo; note the failure on https://github.com/abhinav92003/dynamorio/pull/1, but not on the corresponding merge commit. @derekbruening this is probably because clang-format looks only at the changed lines wrt base branch?

- The automatic emails sent by Github Actions (the ones titled "[DynamoRIO/dynamorio] PR run failed: ci-xxx") has dynamorio@noreply.github.com as its recipient. But I don't see such emails for any PR other than my own (@derekbruening can you please confirm?) . So, there will be duplication of failure  notif email for the PR author.

File: .github/workflows/ci-aarchxx.yml (hunk 5 above)
Comment (derekbruening) [reply .github/workflows/ci-aarchxx.yml:84]: They can't be renamed.  I created new ones with the new names and deleted the old ones.

**Comment by derekbruening** [comment]
> * clang-format doesn't seem to fail after the PR is pushed to master, even though it did fail on the PR. Found this while testing on my forked repo; note the failure on [abhinav92003#1](https://github.com/abhinav92003/dynamorio/pull/1), but not on the corresponding merge commit. @derekbruening this is probably because clang-format looks only at the changed lines wrt base branch?

Sounds right -- yes, it does a "git diff -U0 origin/master".

> * The automatic emails sent by Github Actions to individuals (the ones titled "[DynamoRIO/dynamorio] PR run failed: ci-xxx") has [dynamorio@noreply.github.com](mailto:dynamorio@noreply.github.com) as its recipient. But I don't see such emails for any PR other than my own (@derekbruening can you please confirm?) . So, there will be duplication of failure notif email for the PR author if they are subscribed to dynamorio-devs too.

My "PR run failed" and "Run failed" (manually triggered) emails are `from: Derek Bruening <notifications@github.com>`, to the same recipient you cite, and cc-ed to `ci_activity@noreply.github.com`.  Like you I have not received any from other people's PR's.  Duplication seems fine: we're talking about merges to master, not PR failures, so they should be relatively rare.

**Review by derekbruening — APPROVED** [review]

**Comment by abhinav92003** [comment]
Verified using a simple cron config that it also has `github.ref == 'refs/heads/master'`.

**Comment by abhinav92003** [comment]
run arm tests

**Pull request merged**